src/index.ts

```diff
@@ -28,62 +42,109 @@ export const defaultCompileOptions = (outputOptions: OutputOptions): CompileOpti
+ * @param context Plugin context to bind for each transform instance.
+ * @param options Rollup input options
+ * @param id Rollup's id entry for this source.
+ * @return Instantiated transform class instances for the given entry point.
+ */
+export const instantiateTransforms = (context: PluginContext): Array<Transform> => {
+  return [new IifeTransform(context), new ExportTransform(context), new StrictTransform(context)];
 };
 
-export default function closureCompiler(compileOptions: CompileOptions = {}): Plugin {
-  return {
-    name: 'closure-compiler',
-    transformBundle: (
-      code: string,
-      outputOptions: OutputOptions,
-    ): Promise<{ code: string; map: RawSourceMap } | void> => {
-      const temp = {
-        js: sync(code),
-        map: sync(''),
-      };
+/**
+ * Transform the tree-shaken code from Rollup with Closure Compiler (with derived configuration and transforms)
+ * @param compileOptions Closure Compiler compilation options from Rollup configuration.
+ * @param transforms Transforms to apply to source followin Closure Compiler completion.
+ * @param code Source to compile.
+ * @param outputOptions Rollup Output Options.
+ * @return Closure Compiled form of the Rollup Chunk
+ */
+export const transformChunk = async (
+  transforms: Array<Transform>,
+  compileOptions: CompileOptions,
+  code: string,
+  outputOptions: OutputOptions,
+): Promise<{ code: string; map: RawSourceMap } | void> => {
+  // Each transform has a 'preCompilation' step that must complete before passing
+  // the resulting code to Closure Compiler.
+  for (const transform of transforms) {
+    transform.outputOptions = outputOptions;
+    const result = await transform.preCompilation(code, 'none');
+    if (result && result.code) {
+      code = result.code;
+    }
+  }
+
+  const jsAsFile = sync(code);
+  const mapAsFile = sync('');
 
-      compileOptions = Object.assign(defaultCompileOptions(outputOptions), compileOptions, {
-        js: temp.js,
-        create_source_map: temp.map,
-      });
+  // Compile Options is the final configuration to pass into Closure Compiler.
+  // defaultCompileOptions are overrideable by ones passed in directly to the plugin
+  // but the js source and sourcemap are not overrideable, since this would break the output if passed.
+  compileOptions = {
+    ...defaultCompileOptions(transforms, outputOptions),
+    ...compileOptions,
+    ...{
+      js: jsAsFile,
+      create_source_map: mapAsFile,
+    },
+  };
 
-      const compile: Promise<string> = new Promise((resolve, reject) => {
-        new compiler(compileOptions).run((exitCode: number, stdOut: string, stdErr: string) => {
-          if (exitCode !== 0) {
-            reject(new Error(`Google Closure Compiler exit ${exitCode}: ${stdErr}`));
-          } else {
-            resolve(stdOut);
+  const compile: Promise<string> = new Promise((resolve: (stdOut: string) => void, reject: (error: any) => void) => {
```

Comment:
can we make it a separate function instead of an inner function?

Reply:
Addressed with bf917d4.